### PR TITLE
Getting Started 2.0 - initial draft

### PR DESCRIPTION
**Pull request by rofe**
WIP for #25 

I took a first stab at writing down the progressive disclosure story we discussed at the last hackathon: from simple index.md to custom styling to custom code etc.

Before taking this further, I wanted to get people's opinions to see if it's generally going in he right direction, or if we should try a different approach.

Click [here](https://github.com/adobe/helix-home/blob/getting-started-2.0/getting-started-2.0.md) for the rendered version.

Feedback welcome :)

### CODE REVIEW

**Comment by trieloff** [comment]
I agree on the forking issue. A better way would be to make the instructions explicit "create a new repo, add file with following code, etc" and then point to a repository that you can copy from if you are lost.

I'd also include screenshots of every step.

**Comment by rofe** [comment]
Thanks for your input. I agree with everything. Here's my take on how to advance:
- *Forking:* I agree it's not very helpful, especially with the "progressive disclosure" story. And even with https://github.com/adobe/helix-cli/issues/933 implemented, the user would end up having a whole bunch of example repos lying around after following all the steps. It would indeed be better to start with a single empty repo, and then gradually enhance it to enable more features. Making the user manually copy/paste code from a GitHub repo in the process is the lesser evil, I guess.
- *Personas:* @davidnuescheler could you help come up with a list and description of personas we want to serve?
- *Screenshots:* Absolutely. I actually even started out interspersing a couple, but then decided to remove them again. we should first agree on the right flow before starting this tedious work :)

**Comment by rofe** [comment]
@bdelacretaz would be great to get your input here as well

**Comment by trieloff** [comment]
> Personas

That's a trigger word for me, but here you go:

> Liz is a 23-year old web developer from St. Louis. After finishing her college degree she took a 1-month coding bootcamp in Frontend Web Development from [Savvy Coders covering HTML, CSS, JS, DOM, version control with git and GitHub, and more](https://savvycoders.com/savvy-coders-curriculum/). Following the boot camp, she has VS Code, Chrome and a git client installed on her laptop, but her `node` version is outdated (which she doesn't know). Her GitHub profile has a couple of (mostly empty) example projects as leftovers from the bootcamp courses. She's now looking to build a small portfolio website in order to find a job in web development. Liz has a dog, a black Labrador, who responds to the name "Rat".

**Comment by bdelacretaz** [comment]
I'm not sure if we need such a guide as prose. Different people will look at the examples in different orders, so we might just give an overview of the examples, without being too specific, and point to the collection of examples where people can pick and choose.

We might start with just an overview of what the examples demonstrate, with no details, basically just a list of features that we want to demonstrate in the examples but no instructions.

That can include links to the individual example repositories, but hidden behind keywords, not prominent to make it clear that it's just an overview of the examples.

And then point to [Example Zero](https://github.com/adobe/helix-home/issues/27) which is the only example that we recommend everybody to look at. After that, a [GitHub query](https://github.com/topics/helix-example) lists the other examples so that users can make up their own way among them.

WDYT?

**Comment by rofe** [comment]
> I'm not sure if we need such a guide as prose

I'm on the fence here. I think it's still a valuable exercise to walk through the developer onboarding, even if it's just for the team to wrap their heads around the _progressive exposure_ experience we want to provide. we might for example detect inconsistencies along the way, e.g. step x only makes sense if it comes after step y, not before...

> point to Example Zero (#27)

I agree that having a single official example repo is the way to go. My next iteration will respect that.

**Comment by bdelacretaz** [comment]
> I agree that having a single official example repo is the way to go...

You mean a single repo for all the examples?

I don't think that's good - IIUC the Helix Pages model calls for one repo per website, which naturally leads to one repo per example.

Also, based on my experience with Sling, examples tend to become stale or outmoded and it's good to be able to select exactly which ones we preferably expose, while maybe keeping others around for special cases. 

Having one repo per example allows us to use Git Topics to mark them as current, obsolete etc.

Also, having one big example repo would mask the simplicity of Helix, compared to the initial examples having just a few content files in them to demonstrate simplicity.

In summary, I am strongly in favor of one Git repo per example, using Git topics to make them discoverable (by humans _and_ by the `hlx` command).

**Comment by rofe** [comment]
I just committed an updated flow, trying to incorporate all feedback received so far:
- Forking no more: It now starts from an empty repo in step 1 and keeps evolving it step by step.
- Yay, screenshots: I started including them where I saw fit. Dummies can be replaced with real ones once the flow is less... fluid.
- Single vs multiple example repos: The links to example files currently all point to a single (dummy) repo, but they could easily be spread across multiple example repos. I think that discussion should best be kept in https://github.com/adobe/helix-pages/issues/14 (which I'm still unsure if it shouldn't be moved here from `helix-pages`...)

Please [review again](https://github.com/adobe/helix-home/blob/getting-started-2.0/getting-started-2.0.md) and as always, provide candid feedback :)

File: getting-started-2.0.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+   <br>![](./dummy.png "TODO: Create an index.md")
+1. Add some great Markdown content. Feel free to use [this example](https://raw.githubusercontent.com/adobe/helix-example/master/index.md "example index.md") for starters.
+1. Click the _Commit changes_ button.
+1. Go to `https://<my-repo>-<my_org>.project-helix.page/` to admire your work. Replace `<my-repo>` and `<my-org>` with your actual repository and organization names:
+   <br>![](./dummy.png "TODO: Helix Pages rendering your content")
+1. **Congratulations!** 🎉 You have just published your first page using _Helix Pages_!
+
+> **Hint:** You can add as many Markdown files as you like, and link them in Markdown.
+
+> **Note:** If your GitHub organization contains one or more dashes, add an extra dash between the repository and organization name in the _Helix Pages_ URL. For example, if your organization is called `my-org`, and your repository is `test`, the corresponding _Helix Pages_ URL would be `https://test--my-org.project-helix.page/`.
```
Comment (tripodsan) [inline getting-started-2.0.md]: I would clarify that _github organization_ and `gihub username` means the same here. over even better, don't use _organization_ as we target rather individuals as orgs, I guess.

File: getting-started-2.0.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+   <br>![TODO: Helix Pages rendering your header and footer](./dummy.png)
+
+## Step 3: Add your own HTML files
+
+> You'll need:
+> - basic HTML knowledge
+
+If you have existing HTML files or snippets that you wish to reuse, you can simply add them to your GitHub repository and have _Helix_Pages_ merge and deliver them for you:
+
+1. Add a `static.html` file to your GitHub repository. The file name can be anything, just make sure it has a `.html` extension.
```
Comment (tripodsan) [inline getting-started-2.0.md]: please don't use the name `static`. it has too many connotations already. what about `about.html` ?

File: getting-started-2.0.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+1. Add your HTML. Maybe [like so](https://raw.githubusercontent.com/adobe/helix-example/master/htdocs/static.html "example HTML file")?
```
Comment (tripodsan) [inline getting-started-2.0.md]: same as above ^^^

File: getting-started-2.0.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+1. Click _Commit changes_.
+1. Refresh `https://<my-repo>-<my_org>.project-helix.page/static.html` to see the rendered result.
+
+> **Note:** If there is a `.md` and `.html` with the same name, _Helix Pages_ will give precendence to the Markdown file. Rename or remove the Markdown file to have the HTML file rendered instead.
```
Comment (tripodsan) [inline getting-started-2.0.md]: this is not true. (or will not be true, soon). also, the sentence sounds a bit clumsy / complicated.

File: getting-started-2.0.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+> **Hint:** If you add a static `header.html` or `footer.html` instead of their Markdown versions, _Helix Pages_ will embed these for you.
```
Comment (tripodsan) [inline getting-started-2.0.md]: avoid _static_

File: getting-started-2.0.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+1. Add a `style.css` file to your GitHub repository.
+1. Insert your CSS rules. See the [default CSS](https://raw.githubusercontent.com/adobe/helix-pages/master/htdocs/style.css "default CSS") for pointers where you can add overrides.
+1. Click _Commit changes_.
+1. Reload `https://<my-repo>-<my_org>.project-helix.page/` to view your style changes in action.
+1. Repeat steps 2 - 4 until you are happy with how your page looks.
+
+### Local development
+
+For an improved developer experience, using a local environment and the IDE of your choice, we proudly introduce you to the _Helix CLI_ (`hlx`). For the sake of simplicity, we just assume you are using _Visual Studio Code_ and have a `code` alias for it.
```
Comment (tripodsan) [inline getting-started-2.0.md]: why do you need _Visual Code_ ? I don't think this is a requirement. at least not until you debug code....

File: getting-started-2.0.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+> You'll need:
+> - a shell :)
+> - an IDE, e.g. [Visual Studio Code](https://code.visualstudio.com/)
```
Comment (tripodsan) [inline getting-started-2.0.md]: IDE is not required.

```suggestion
> - an Text editor or IDE, suitable to edit markdown, html and css files.
```

File: getting-started-2.0.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+> - [Node.js](https://nodejs.org/en/) 8.9 or higher and `npm`
```
Comment (tripodsan) [inline getting-started-2.0.md]: - `npm` or `yarn`

File: getting-started-2.0.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+1. Click _Commit changes_.
+1. Reload `https://<my-repo>-<my_org>.project-helix.page/` to view your style changes in action.
+1. Repeat steps 2 - 4 until you are happy with how your page looks.
+
+### Local development
+
+For an improved developer experience, using a local environment and the IDE of your choice, we proudly introduce you to the _Helix CLI_ (`hlx`). For the sake of simplicity, we just assume you are using _Visual Studio Code_ and have a `code` alias for it.
+
+> You'll need:
+> - a shell :)
```
Comment (tripodsan) [inline getting-started-2.0.md]: what is a shell? :shell: ?

- a command line terminal (linux, macos) or command prompt (windows)

but since every OS has some sort of terminal/command prompt, it is not really a requirement.

File: getting-started-2.0.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+> - an IDE, e.g. [Visual Studio Code](https://code.visualstudio.com/)
+> - [Node.js](https://nodejs.org/en/) 8.9 or higher and `npm`
+> - [Helix CLI](https://www.npmjs.com/package/@adobe/helix-cli)
+> - [Git](https://git-scm.com/)
+
+1. Open your favorite shell.
```
Comment (tripodsan) [inline getting-started-2.0.md]:
```suggestion
1. Open your favorite command line terminal.
```

File: getting-started-2.0.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+1. Change to the directory where you normally keep your code, e.g. type `cd ~/code`.
+1. Type `git clone https://github.com/<my-org>/<my-repo>.git`
+   <br>This will your GitHub repository locally into a folder with the same name as your GitHub repository.
```
Comment (tripodsan) [inline getting-started-2.0.md]: - explain how to use the `clone` button form git.
- many people don't use the command line tools to operate git. but use the git integration of they IDE or a different git client...so I wonder if we should add a note regarding this:

> **Note**: This guide uses the git command line tools to operate git, but this is not a requirement. Feel free to use a git client of your choice.

File: getting-started-2.0.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+1. Type `cd <my_repo>`
+1. Type `hlx up &`
```
Comment (tripodsan) [inline getting-started-2.0.md]: why `&` ? (doesn't work on windows)

File: getting-started-2.0.md [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+   <br>This will build a _Helix Pages_ project locally, start a local web server in the background and open a new browser window at [`http://localhost:3000/`](http://localhost:3000/ "your local Helix server") when done.
+1. View your page rendered locally by _Helix Pages_ using the default page styling.
+1. Type `code .` to open this project in the IDE
```
Comment (tripodsan) [inline getting-started-2.0.md]: why? as mentioned above, I would not rely on an existing IDE.

File: getting-started-2.0.md [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+1. Create a `style.css` file and insert your CSS rules. See the [default CSS](https://raw.githubusercontent.com/adobe/helix-pages/master/htdocs/style.css "Default CSS") for pointers where you can add overrides.
```
Comment (tripodsan) [inline getting-started-2.0.md]:
```suggestion
1. Create a `style.css` file with your favorite text editor or IDE and insert your CSS rules. See the [default CSS](https://raw.githubusercontent.com/adobe/helix-pages/master/htdocs/style.css "Default CSS") for pointers where you can add overrides.
```

File: getting-started-2.0.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+## Add your own code
+
+> You'll need:
+> - basic JavaScript knowledge
+> - basic HTML knowledge
+
+_Helix Pages_ allows you to control the rendering of your content by changing the HTL template and adding custom JavaScript code.
+
+1. Open your favorite shell and change to the directory where you checked out your GitHub repository, e.g. type `cd ~/code/<my-repo>`.
+2. Type `code .` to open this project in the IDE.
```
Comment (tripodsan) [inline getting-started-2.0.md]: same as above... don't favour visual code

File: getting-started-2.0.md [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+</div>
+```
+9. Save the changes.
+10. Reload [`http://localhost:3000/`](http://localhost:3000/ "your local Helix server") to see the page rendered locally using your custom code.
+11. Type `git add src/* && git commit -m"custom code" && git push`
+    <br>This will commit your local code changes to the GitHub repository.
+12. Go to `https://<my-repo>-<my_org>.project-helix.page/` and observe the published page being rendered with your custom code. 
+
+
+> **Note:** You could also add custom code directly in GitHub, but we recommend against writing code outside a proper development environment.
```
Comment (tripodsan) [inline getting-started-2.0.md]: why is this note neccesairy?

File: getting-started-2.0.md [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+## Add your own domain
+
+> You'll need
+> - A [Fastly](https://github.com/adobe/project-helix/blob/master/SERVICES.md#fastly) account
```
Comment (tripodsan) [inline getting-started-2.0.md]: no. you don't :-)

File: getting-started-2.0.md (hunk 4 above)
Comment (rofe) [reply getting-started-2.0.md]: Oh, the I misunderstood https://github.com/adobe/helix-publish/issues/96 :) will it be the other way round?

File: getting-started-2.0.md (hunk 6 above)
Comment (rofe) [reply getting-started-2.0.md]: It doesn't say you _need_ it :) it is just an assumption.

File: getting-started-2.0.md (hunk 9 above)
Comment (rofe) [reply getting-started-2.0.md]: Here I thought of the thing running _inside_ the terminal and executing your commands, i.e. `bash` or `zsh` or... but you're right, it doesn't matter which you use, and every OS provides one. The key is to find the terminal ;)

File: getting-started-2.0.md (hunk 12 above)
Comment (rofe) [reply getting-started-2.0.md]: Because it always happens to me that `hlx` is the front-most process and if I want to type `code .` next, I either have to open a new terminal tab or suspend and throw it in the background first :)

Open to a generic solution that would work for all OSs...

File: getting-started-2.0.md (hunk 16 above)
Comment (rofe) [reply getting-started-2.0.md]: I wanted to point out that having a local dev env isn't strictly mandatory to work with Helix Pages. Would you rephrase it or remove it?

File: getting-started-2.0.md (hunk 17 above)
Comment (rofe) [reply getting-started-2.0.md]: Got me! I was guessing :)

File: getting-started-2.0.md (hunk 4 above)
Comment (tripodsan) [reply getting-started-2.0.md]: yes. `.html` takes precedence.

File: getting-started-2.0.md [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+
+The easiest way to publish content is to have [_Helix Pages_](https://www.project-helix.page/) render and deliver a Markdown file in your GitHub repository:
+
+1. Click the _Create new repository_ button to set up a new public [GitHub repository](https://github.com/new/).
+   <br>![](./dummy.png "TODO: Create a repository")
+1. Click the _Create new file_ button in your GitHub repository to add a new file called `index.md`. This will be your homepage.
+   <br>![](./dummy.png "TODO: Create an index.md")
+1. Add some great Markdown content. Feel free to use [this example](https://raw.githubusercontent.com/adobe/helix-example/master/index.md "example index.md") for starters.
+1. Click the _Commit changes_ button.
+1. Go to `https://<my-repo>-<my_org>.project-helix.page/` to admire your work. Replace `<my-repo>` and `<my-org>` with your actual repository and organization names:
```
Comment (tripodsan) [inline getting-started-2.0.md]: use `<my-username>` instead of the org.

File: getting-started-2.0.md (hunk 12 above)
Comment (tripodsan) [reply getting-started-2.0.md]: but this is a problem of the user. if the user is savy, it will know to use `&` or a second terminal. If not, well...

you can also tell him to do:
```
ctrl-z
$ open .
$ fg
```

but this would also be very platform specific.

File: getting-started-2.0.md (hunk 4 above)
Comment (rofe) [reply getting-started-2.0.md]: Thanks. Changed it.

File: getting-started-2.0.md (hunk 12 above)
Comment (rofe) [reply getting-started-2.0.md]: Ok, you're right. I'll remove it.

Comment (tripodsan) [reply getting-started-2.0.md]: also, you'd need to explain how to stop the server....

**Comment by rofe** [comment]
Great feedback @tripodsan! The new revision is available now.

More reviews?

File: getting-started-2.0.md [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+## Add your own styling
+
+> You'll need:
+> - basic CSS knowledge
+
+_Helix Pages_ allows you to override the default CSS. You can either do this directly in GitHub, or start developing locally.
+
+### GitHub
+
+1. Add a `style.css` file to your GitHub repository.
```
Comment (trieloff) [inline getting-started-2.0.md]: Does this have to go into `htdocs`?

Comment (rofe) [reply getting-started-2.0.md]: No: I thought the idea was to overlay from the "content" repo, while `htdocs` is something for the "code" repo.

File: getting-started-2.0.md [hunk 20]
```diff
@@ -0,0 +1,152 @@
+# Getting Started with Helix
+
+This guide helps you as a web developer get started with Helix with very little effort, then progressively discover more features as you need them.
+
+## Step 1: Publish your content (look ma, no hands!)
```
Comment (bdelacretaz) [inline getting-started-2.0.md]: For someone who's familiar with GitHub, I think this can be simplified to
> Create a public GitHub repository with an `index.md` file in it and open to `https://<my-repo>-<my-name>.project-helix.page/` to enjoy the results. Next (or instead of creating your own repository) have a look at https://github.com/bdelacretaz/helix-example-zero which is similar to that, with headers/footers, CSS and a few other self-explaining features.

I understand the desire to explain the steps for someone who's _not_ familiar with GitHub, but for someone who is it feels like kindergarten ;-)

Not sure how best to fix this - maybe add the above short explanation at the top of this section under a "if you're familiar with GitHub and Markdown" heading?

Comment (bdelacretaz) [reply getting-started-2.0.md]: Or maybe we point the people who are familiar with GitHub directly to (and improved version of) https://github.com/bdelacretaz/helix-example-zero and this document is for people who are not.

Comment (bdelacretaz) [reply getting-started-2.0.md]: I have clarified the above ideas in the "Three main personas for documentation?" section at https://github.com/adobe/helix-home/issues/31

**Comment by bdelacretaz** [comment]
I have commented on the changes and later realized that what I'm really after is the "Three main personas for documentation?" section at #31. Agreeing on those (or different) personas would help us decide where pure examples make sense, vs. guides written in prose.

**Comment by rofe** [comment]
Good point @bdelacretaz. Maybe this document should be renamed to _helix-for-dummies.md_ then? ;)

**Comment by bdelacretaz** [comment]
> Maybe this document should be renamed to helix-for-dummies.md then? 

Maybe ;-)

Or we might introduce this guide as follows:

* If you are familiar with GitHub, Markdown, HTML and CSS you may want to proceed directly to the [URL of the "pure content" example]
* If in addition to that you are familiar with Node.js programming you may want to proceed directly to the [URL of the "programming" example]
* If you need help understanding some of these tools and concepts, we recommend that you start with the step-by-step tutorial below.
* In any case, our reference documentation at [URL of the docs] provides all the details that you need, once you grasp the basic concepts.

So that people can choose between a fast track looking at the examples and the step-by-step guide.

**Comment by rofe** [comment]
Incorporated @bdelacretaz's feedback in the [latest version](https://github.com/adobe/helix-home/blob/getting-started-2.0/getting-started-2.0.md).

File: getting-started-2.0.md [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+1. Click the _Commit changes_ button.
+1. Reload `https://<my-repo>-<my-name>.project-helix.page/` to see header and footer in action.  
+![TODO: Helix Pages rendering your header and footer](./dummy.png)
+
+## Step 3: Add your own HTML files
+
+> You'll need:
+> - basic HTML knowledge
+
+If you have existing HTML files or snippets that you wish to reuse, you can simply add them to your GitHub repository and have _Helix_Pages_ merge and deliver them for you:
```
Comment (stefan-guggisberg) [inline getting-started-2.0.md]: _Helix_Pages_ => _Helix Pages_

File: getting-started-2.0.md [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+1. Add a logo, and maybe a list of other links you would like to show in the header navigation. Here's some [inspiration](https://raw.githubusercontent.com/adobe/helix-example-first/master/header.md "example header.md").
+1. Click the _Commit changes_ button.
+1. If you added a logo in step 2, upload that now.  
+![](./dummy.png "TODO: Upload logo")
+1. Add a file called `footer.md` to your GitHub repository.
+1. Add some content you would like to be shown in the footer. Here's a [proposal](https://raw.githubusercontent.com/adobe/helix-example-first/master/footer.md "example footer.md").
+1. Click the _Commit changes_ button.
+1. Reload `https://<my-repo>-<my-name>.project-helix.page/` to see header and footer in action.  
+![TODO: Helix Pages rendering your header and footer](./dummy.png)
+
```
Comment (kptdobe) [inline getting-started-2.0.md]: I would insert a new step 3: Use a branch

```suggestion
## Step 3: Use a branch

By default, _Helix Pages_ renders the `master` branch of your GitHub repository. It can also render any branch of your repository using the pattern `https://<my-branch>--<my-repo>--<my-name>.project-helix.page/` (note the two dashes, they are required).

1. Create a branch.
![TODO: Create a branch.md](./dummy.png)
1. Edit the `index.md` file in the new branch.
![TODO: Edit index.md in the branch](./dummy.png)
1. Click the _Commit changes_ button.
1. Load https://<my-branch>--<my-repo>--<my-name>.project-helix.page/
![TODO: Helix Pages rendering your branch](./dummy.png)
```

Comment (kptdobe) [reply getting-started-2.0.md]: Naturally, this shifts all the numbers in the titles :)

Comment (rofe) [reply getting-started-2.0.md]: Good idea! But how about introducing it at step 2, where you still only focus on `index.md` and now get to play with variants?

Comment (kptdobe) [reply getting-started-2.0.md]: That was my initial thought but I realised that this is a more "advanced" concept. header/footer are probably more of interest for a beginner.

Comment (rofe) [reply getting-started-2.0.md]: I just played around a bit with separate step, substep of step 1, and making it a hint at the end of step 1, and I think I prefer the latter: it keeps the tutorial flow intact without introducing too much complexity and extra steps with screenshots etc, and contains a link to details about branching for those who are interested.

Let me know what you think!
